src/axom/core/Array.hpp

```diff
@@ -550,121 +830,303 @@ Array< T >::~Array()
 }
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::insert( const T& value, IndexType pos )
+template< typename ... Args >
+inline void Array< T >::emplace( IndexType pos, Args&&... args )
 {
-  assert( m_num_components == 1 );
   reserveForInsert( 1, pos );
-  m_data[ pos ] = value;
+  m_data[ pos ] = std::move( T(std::forward<Args>(args)...));
 }
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::insert( const T* tuples, IndexType n, IndexType pos )
+template< typename ... Args >
+inline typename Array< T >::ArrayIterator
+Array< T >::emplace( Array< T >::ArrayIterator pos, Args&&... args )
 {
-  assert( tuples != nullptr );
-  T* insert_pos = reserveForInsert( n, pos );
-  std::memcpy( insert_pos, tuples, n * m_num_components * sizeof(T) );
+  assert( pos >= begin() && pos <= end() );
+  emplace( pos - begin(), args... );
+  return pos;
 }
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::emplace( IndexType n, IndexType pos, const T& value )
+template< typename ... Args >
+inline void Array< T >::emplace_back( Args&&... args )
 {
-  T* insert_pos = reserveForInsert( n, pos );
-  std::fill_n( insert_pos, n * numComponents(), value );
+  emplace( m_num_elements, args... );
 }
 
+//------------------------------------------------------------------------------
+template< typename T >
+inline void Array< T >::resize( IndexType new_num_elements )
+{
+  assert( new_num_elements >= 0 );
+
+  if ( new_num_elements > m_capacity )
+  {
+    dynamicRealloc( new_num_elements );
+  }
+
+  updateNumElements( new_num_elements );
+}
+
+//------------------------------------------------------------------------------
+template< typename T >
+inline void Array< T >::swap ( Array< T >& other )
+{
+  T* temp_data = m_data;
+  IndexType temp_num_elements = m_num_elements;
+  IndexType temp_capacity = m_capacity;
+  double temp_resize_ratio = m_resize_ratio;
+  bool temp_is_external = m_is_external;
+
+  m_data = other.m_data;
+  m_num_elements = other.m_num_elements;
+  m_capacity = other.m_capacity;
+  m_resize_ratio = other.m_resize_ratio;
+  m_is_external = other.m_is_external;
+
+  other.m_data = temp_data;
+  other.m_num_elements = temp_num_elements;
+  other.m_capacity = temp_capacity;
+  other.m_resize_ratio = temp_resize_ratio;
+  other.m_is_external = temp_is_external;
+
+}
 
 //------------------------------------------------------------------------------
 template< typename T >
-inline void Array< T >::resize( IndexType new_num_tuples )
+inline std::ostream& Array< T >::print(std::ostream& os) const
 {
-  assert( new_num_tuples >= 0 );
 
-  if ( new_num_tuples > m_capacity )
+  #if defined(AXOM_USE_UMPIRE)
+    if (m_allocator_id == 
+          axom::getUmpireResourceAllocatorID( umpire::resource::Device ) ||
```

Comment:
I don't think this check is sufficient here. 

An application may have created a custom allocator, e.g., a pool allocator, on the Device that has a different ``allocator_id``, but is on the Device.

 We should check what else is available through the Umpire API. Specifically, given an ``allocator_id``, is there a way to query the Umpire Resource type associated with a given allocator?